### ISSUE TITLE
Question: Is it possible to order by apply-function result?

### ISSUE THREAD

**Issue by tariqalardah** [redis/redis-om-dotnet#48]
I am trying to order items by apply-function result, like:
```
var outletsAggregationSet = provider.AggregationSet<Outlet>();

var result = outletsAggregationSet
            .Apply(x => ApplyFunctions.GeoDistance(x.RecordShell.Location, myLon, myLat), "Distance")
            .OrderBy(/*How to order by Distance?*/);

```
Is there anyway I can sort the results by calculated "Distance" as done in the following command:
`FT.AGGREGATE outlets-index '@Name:baskin @Location:[55.397 25.123 10 km]' LOAD 2 @Name @Location  APPLY 'geodistance(@Location,"55.397 25.123")' AS distance SORTBY 2 @distance ASC`

**Comment by slorello89** [comment]
```
var outletsAggregationSet = provider.AggregationSet<Outlet>();

var result = outletsAggregationSet
            .Apply(x => ApplyFunctions.GeoDistance(x.RecordShell.Location, myLon, myLat), "Distance")
            .OrderBy(x=>x["Distance"]);

```

should work.

**Comment by tariqalardah** [comment]
> ```
> var outletsAggregationSet = provider.AggregationSet<Outlet>();
> 
> var result = outletsAggregationSet
>             .Apply(x => ApplyFunctions.GeoDistance(x.RecordShell.Location, myLon, myLat), "Distance")
>             .OrderBy(x=>x["Distance"]);
> ```
> 
> should work.

It works, thank you
But, it is hard for me to debug the issues I am facing:
Adding `.Where(x => x.RecordShell.Name == "baskin")` before `Apply()` 
debugging args are `["outletsindex","","APPLY","geodistance(@Location,55.3974726,25.1232973)","AS","Distance","SORTBY","2","@Distance","ASC"]) ` Note the second is empty string
results in 0 records 
While adding `.Where(x => x.RecordShell.Name == "baskin")` after `Apply()`
debugging args are 
`["outlets-index","*","APPLY","geodistance(@Location,55.3974726,25.1232973)","AS","Distance","FILTER","@Name == baskin","SORTBY","2","@Distance","ASC"]`
Note now the second is "*"
And results an exception : `{"Missing (or badly formatted) value for expr->pred.right"}`
   ```
at StackExchange.Redis.ConnectionMultiplexer.ExecuteSyncImpl[T](Message message, ResultProcessor`1 processor, ServerEndPoint server) in /_/src/StackExchange.Redis/ConnectionMultiplexer.cs:line 2886
   at StackExchange.Redis.RedisBase.ExecuteSync[T](Message message, ResultProcessor`1 processor, ServerEndPoint server) in /_/src/StackExchange.Redis/RedisBase.cs:line 54
   at StackExchange.Redis.RedisDatabase.Execute(String command, ICollection`1 args, CommandFlags flags) in /_/src/StackExchange.Redis/RedisDatabase.cs:line 1227
   at StackExchange.Redis.RedisDatabase.Execute(String command, Object[] args) in /_/src/StackExchange.Redis/RedisDatabase.cs:line 1223
   at Redis.OM.RedisConnection.Execute(String command, String[] args) in ...\redis-om-dotnet\src\Redis.OM\RedisConnection.cs:line 27
   at Redis.OM.Aggregation.AggregationEnumerator`1.StartEnumeration() in ...\redis-om-dotnet\src\Redis.OM\Aggregation\AggregationEnumerator.cs:line 212
   at Redis.OM.Aggregation.AggregationEnumerator`1.MoveNext() in ...\redis-om-dotnet\src\Redis.OM\Aggregation\AggregationEnumerator.cs:line 134
```

Not sure, if it is expected, or should file a bug with these details!

**Comment by slorello89** [comment]
Hi @tariqalardah can you share your `Outlet` class here? I would think the first way at least should work as that's going to the main query.

Interestingly I found that if it's string equality in the FILTER predicate, the expression Parser is not wrapping the strings with single quotes which is causing the exact error you are referring to, clearly that's a bug, you can work around it for the moment by wrapping the query with `'` yourself when the where clause is after the apply clause. e.g. 

```csharp
.Where(x => x.RecordShell.Name == "'baskin'")
```

If you could share the Outlet class with me though that would be helpful, I had no problem when I had the `Where` clause first, it was properly building the query for me.

thanks!

**Comment by tariqalardah** [comment]
Hi @slorello89, thank you for the response
Here is the outlet class:
```
[Document(IndexName = "outlets-index")]
public class Outlet
{
    [RedisIdField]
    public int Id { get; set; }
    [Searchable(Aggregatable = true)]
    public string Name { get; set; }
    [Indexed] 
    public string Merchant { get; set; }
    [Indexed]
    public string Logo { get; set; }
    [Indexed(Aggregatable = true)]
    public GeoLoc Location { get; set; }
}
```
Wrapping the filter term with '' eliminated the exception, but I am still trying to figure out why it is returning 0 results (it is not being evaluated correctly).
Another question please, is there a way to override the general query defaulted to `"*"` instead of adding filter when working with `RedisAggregationSet`?

**Comment by slorello89** [comment]
Hi @tariqalardah, had a chance to look into this a bit more.

If you put the `Where` clause before anything else it will automatically apply that clause as the query, e.g.:

```csharp
foreach(var result in set
            .Where(x=>x.RecordShell.Name == "baskin")
            .Apply(x=>ApplyFunctions.GeoDistance(x.RecordShell.Location, loc), "distance")
            .OrderBy(x=>x["distance"]))
```
yields:
```
1643321051.932273 [0 172.17.0.1:51242] "FT.AGGREGATE" "outlets-index" "@Name:baskin" "APPLY" "geodistance(@Location,55.3974726,25.1232973)" "AS" "distance" "SORTBY" "2" "@distance" "ASC"
```

2. As to the reason you aren't getting results, can you run an `FT.INFO`. I would guess one of two things are happening

1: You are not setting an `Id` for the objects you are inserting into Redis, if you do not use a `Guid` or `Ulid`, Redis.OM will NOT attempt to generate an Id for you and you will need to register an `IIdGenerationStrategey` with the `DocumentAttribute`. Because no one is setting your ids they are falling back to the integer default of 0 for everything so only a single record is being indexed

2: There's some odd indexing failure occurring when you insert objects into Redis, if you see any indexing failures please share an example of some of the objects stored in Redis so I can take a look.

**Comment by tariqalardah** [comment]
Hi @slorello89,
Thank you for the time you put into this.

Regarding issues in indexing, I don't think there are issues, as the integer id is being used. 
Ex (`keys *`):
```
   1) "RedisTests.Outlet:3192"
   2) "RedisTests.Outlet:95106"
   3) "RedisTests.Outlet:57983"
...
```
Here is index info
```
*.cloud.redislabs.com:*> FT.INFO outlets-index
 1) index_name
 2) outlets-index
 3) index_options
 4) (empty array)
 5) index_definition
 6) 1) key_type
    2) HASH
    3) prefixes
    4) 1) RedisTests.RedisOutlet:
    5) default_score
    6) "1"
 7) attributes
 8) 1) 1) identifier
       2) Name
       3) attribute
       4) Name
       5) type
       6) TEXT
       7) WEIGHT
       8) "1"
       9) SORTABLE
    2) 1) identifier
       2) Merchant
       3) attribute
       4) Merchant
       5) type
       6) TAG
       7) SEPARATOR
       8) |
    3) 1) identifier
       2) Logo
       3) attribute
       4) Logo
       5) type
       6) TAG
       7) SEPARATOR
       8) |
    4) 1) identifier
       2) Location
       3) attribute
       4) Location
       5) type
       6) GEO
       7) SORTABLE
 9) num_docs
10) "2654"
11) max_doc_id
12) "2654"
13) num_terms
14) "2642"
15) num_records
16) "24746"
17) inverted_sz_mb
18) "0.12707996368408203"
19) total_inverted_index_blocks
20) "5607"
21) offset_vectors_sz_mb
22) "0.016357421875"
23) doc_table_size_mb
24) "0.277923583984375"
25) sortable_values_size_mb
26) "0.25993251800537109"
27) key_table_size_mb
28) "0.09801483154296875"
29) records_per_doc_avg
30) "9.3240394592285156"
31) bytes_per_record_avg
32) "5.3848299980163574"
33) offsets_per_term_avg
34) "0.69312214851379395"
35) offset_bits_per_record_avg
36) "8"
37) hash_indexing_failures
38) "0"
39) indexing
40) "0"
41) percent_indexed
42) "1"
43) gc_stats
44)  1) bytes_collected
     2) "0"
     3) total_ms_run
     4) "0"
     5) total_cycles
     6) "0"
     7) average_cycle_time_ms
     8) "-nan"
     9) last_run_time_ms
    10) "0"
    11) gc_numeric_trees_missed
    12) "0"
    13) gc_blocks_denied
    14) "0"
45) cursor_stats
46) 1) global_idle
    2) (integer) 0
    3) global_total
    4) (integer) 0
    5) index_capacity
    6) (integer) 128
    7) index_total
    8) (integer) 0
```
I believe there is an issue in Redis.OM handling `Where` at the beginning. Here is an example of query I captured in `redis-cli monitor` where the query is obviously "", and also, I don't know why the following 2 commands:
```
1643470072.187264 [0 *.*.*.*:63872] "FT.AGGREGATE" "outlets-index" "" "APPLY" "geodistance(@Location,55.3974726,25.1232973)" "AS" "Distance" "SORTBY" "2" "@Distance" "ASC"
1643470072.239264 [0 *.*.*.*:63872] "SELECT" "0"
1643470072.239264 [0 *.*.*.*:63872] "FT.CURSOR" "DEL" "outlets-index" "-1"
```

Regarding, why no results while using `Where `after `Apply`, it is my mistake may be as I thought `.Where(x=>x.RecordShell.Name == "baskin")` will work as `@Name:baskin` in the query to result in all names starts with baskin! But apprently it translate to `"FILTER" "@Name == 'baskin'"` and dont work like this, it only return outlet where exact name is 'baskin'

**Comment by slorello89** [comment]
Hey @tariqalardah can you share the exact query you used that produced 

```
1643470072.187264 [0 *.*.*.*:63872] "FT.AGGREGATE" "outlets-index" "" "APPLY" "geodistance(@Location,55.3974726,25.1232973)" "AS" "Distance" "SORTBY" "2" "@Distance" "ASC"
```

So that I can try to trace the issue, for me:

```csharp
foreach(var result in set
            .Apply(x=>ApplyFunctions.GeoDistance(x.RecordShell.Location, loc), "distance")
            .Where(x=>x.RecordShell.Name == "'baskin'")
            .OrderBy(x=>x["distance"]));
```
produces 

```bash
1643632346.505311 [0 172.17.0.1:51876] "FT.AGGREGATE" "outlets-index" "*" "APPLY" "geodistance(@Location,55.3974726,25.1232973)" "AS" "distance" "FILTER" "@Name == 'baskin'" "SORTBY" "2" "@distance" "ASC"
```
and 

```csharp
foreach(var result in set
            .Where(x=>x.RecordShell.Name == "baskin")
            .Apply(x=>ApplyFunctions.GeoDistance(x.RecordShell.Location, loc), "distance")
            .OrderBy(x=>x["distance"]));
```

(Notice I dropped the `'` around baskin) Produces:

```csharp
1643632859.266886 [0 172.17.0.1:51884] "FT.AGGREGATE" "outlets-index" "@Name:'baskin'" "APPLY" "geodistance(@Location,55.3974726,25.1232973)" "AS" "distance" "SORTBY" "2" "@distance" "ASC"
```

I wonder if somehow how the translator is unpacking your query it's just getting an empty string.

As for the `SELECT`/Cursor commands, the `SELECT` is done by the underlying client library to select a database in Redis (0 is the default database), and `FT.CURSOR` is part of the internal pagination engine of Redis.OM.

**Comment by tariqalardah** [comment]
Hey @slorello89,
I finally found where the problem is!

It is in:
`public static RedisAggregationSet<T> Where<T>(this RedisAggregationSet<T> source, Expression<Func<AggregationResult<T>, bool>> expression)`

And all what you need to reproduce it is to only store the search query term in a string variable instead of using constant string in `.Where()`.
.Where(x=>x.RecordShell.Name == "baskin")
In your code snippet try `.Where(x=>x.RecordShell.Name == term)` where `var term = "baskin";`

**Comment by slorello89** [comment]
Hey, @tariqalardah thanks for bearing with me, yep that's definitely it, there is an issue in how it's unpacking the frames in the expression parsers #62 should fix this.

**Comment by slorello89** [comment]
Hi @tariqalardah, this should be fixed by #62 in v0.1.6, which is rolling out now.

**Comment by slorello89** [comment]
@tariqalardah I'm going to close this issue now as it should be resolved, feel free to re-open if you still see the problem. Thanks!
